File: python/2024/day2.py
with open("inputs/day2.txt","rt") as fp:
    reports = [list(map(int,line.split())) for line in fp.readlines()]
   
def isSafe(report):
    dir=0;
    for i in range(len(report)-1):
        diff=report[i]-report[i+1]
        if (diff==0):
            return False
        currdir=diff/abs(diff)
        if (dir==0):
            dir=currdir
        else:
            if (dir!=currdir): 
                return False
        if ((abs(diff) <1) or (abs(diff) >3)):
            return False
    return True
    
safecount = 0
safecount_1 = 0
for report in reports:
    if (isSafe(report)):
        safecount+=1
        continue
    else:
        for i in range(len(report)):
            reducedRep = report.copy()
            reducedRep.pop(i)
            if (isSafe(reducedRep)):
                safecount_1+=1
                break
        
        
        
print("Task1: ",safecount)
print("Task2: ",safecount+safecount_1)

File: Cpp/2024/day2.cpp
#include "../Common/Advent_of_Code.h"


#include <deque>
#include <iostream>
#include <fstream>
#include <filesystem>
#include <cmath>
#include <iterator>
#include <numeric>
#include <sstream>

template <typename T> int sgn(T val) {
    return (T(0) < val) - (val < T(0));
}

bool isSafe(const std::deque<int> &numbers)
{
    std::deque<int> differences;
    std::adjacent_difference(numbers.cbegin(),numbers.cend(),std::back_inserter(differences));
    differences.pop_front();
    return std::none_of(differences.cbegin(),differences.cend(),[](auto &&val){return abs(val)>3 || val==0;}) &&
           std::none_of(std::next(differences.cbegin()),differences.cend(),[&differences](auto value){return sgn(value) != sgn(differences.at(0));});
}

void day2()
{
    //std::ifstream inputFile(std::filesystem::path("../../inputs/day2_training.txt"));
    std::ifstream inputFile(std::filesystem::path("../../inputs/day2.txt"));
    int count = 0;
    int dampenedCount = 0;
    for (std::string line; std::getline(inputFile,line);)
    {
        std::istringstream in(line);
        std::istream_iterator<int> numberIt(in);
        std::deque<int> numbers;
        std::copy(numberIt,std::istream_iterator<int> (),std::back_insert_iterator(numbers));
        if (isSafe(numbers)) count++;
        else{
            for (int i=0;i<numbers.size();i++)
            {
                std::deque<int> numbersOneRemoved=numbers;
                numbersOneRemoved.erase(numbersOneRemoved.begin()+i);
                if (isSafe(numbersOneRemoved)){
                    dampenedCount++;
                    break;
                }
            }
        }
    }
    std::cout << "Day2 task1: " << count << std::endl;
    std::cout << "Day2 task2: " << count+dampenedCount << std::endl;
}
